### PR TITLE
Add preliminary GLSL support

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Part of https://github.com/compiler-explorer/compiler-explorer/issues/2331
Inspired by https://github.com/compiler-explorer/compiler-explorer/pull/3932

I think it is great that HLSL has support here, but the lack of GLSL support is what stops me from using Compiler-Explorer more often. LunarG also have shown in our [yearly ecosystem survey](https://www.lunarg.com/wp-content/uploads/2024/04/2024-Vulkan-Ecosystem-Survey-Results-04-04-2024.pdf) that GLSL is very widely used in the GPU space 
image

A while ago I started to create [my own SPIR-V tool](https://github.com/sjfricke/SPIRV-Playground) but was given feedback to "please just spend your free-time efforts making compiler-explorer better for SPIR-V instead" ... to which I agree (since seems a LOT of ground work was already laid out for me :smile: )

So here it is, a working version of GLSL using `glslang` :rocket: 

image

I guess beside reviewing my very unused Typescript skills, I know we will need to add things to the Infra repo to get things public. For that I am not sure what the best course of action is, we do have [rolling releases](https://github.com/KhronosGroup/glslang/releases/tag/main-tot) of `glslang`  as well as SDK version tags

### CODE REVIEW

File: etc/config/glsl.defaults.properties [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+defaultCompiler=glslangValidator
+supportsBinary=false
+compilerType=glsl
+instructionSet=spirv
+
+disassemblerPath=/opt/compiler-explorer/SPIRV-Tools-master/build/tools/spirv-dis
+
+group.glslang.compilers=glslangValidator
+compiler.glslangValidator.exe=/home/fricke/1.3.290.0/x86_64/bin/glslangValidator
```
Comment (partouf) [inline etc/config/glsl.defaults.properties:11]: Should be something default like when you `cmake --install` for example or install from `apt-get` or similar. (`/usr/bin/glslangValidator` maybe?)

Comment (spencer-lunarg) [reply etc/config/glsl.defaults.properties:11]: I guess not sure if `glslang` has an `apt-get` packaging as I only ever get it from the Vulkan SDK ... (this was clearly for testing, I was hoping someone tell me what it should actually should be for the checked in version, so I guess `/usr/bin/glslangValidator` is better then nothing)

**Comment by spencer-lunarg** [comment]
Hey friendly ping, I know people are busy but I just wanted to know if there is something on my end that I still need to do or if it just a matter of getting time to assess adding new langauges/compilers

**Comment by partouf** [comment]
> Hey friendly ping, I know people are busy but I just wanted to know if there is something on my end that I still need to do or if it just a matter of getting time to assess adding new langauges/compilers

no nothing to do, this is fine, but we'll get to it when we get to it

File: examples/glsl/default.glsl [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+#version 450
+layout(set = 0, binding = 0, rgba8) readonly uniform image2D myImage;
+layout(set = 0, binding = 1, std430) buffer SSBO {
+    ivec2 coords;
+    vec4 data;
+};
+
+void main() {
+    data = imageLoad(myImage, coords);
+}
```
Comment (mattgodbolt) [inline examples/glsl/default.glsl:12]: is the missing newline at the end of the file intentional?

File: static/modes/glsl-mode.ts [hunk 3]
```diff
@@ -0,0 +1,745 @@
+// Copyright (c) 2022, Compiler Explorer Authors
```
Comment (mattgodbolt) [inline static/modes/glsl-mode.ts:1]:
```suggestion
// Copyright (c) 2024, Compiler Explorer Authors
```

File: static/modes/glsl-mode.ts [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+        'local_size_z_id',
+        'precise',
+        'noperspective',
+        'writeonly',
+        'readonly',
+        'in',
+        'out',
+        'inout',
+
+        // scalar types
```
Comment (mattgodbolt) [inline static/modes/glsl-mode.ts:163]: Aren't a few of these already C++ keywords? Is it OK to put duplicates into monaco?

e.g.` const` and `float` at least

File: static/modes/glsl-mode.ts [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+        'usampler2DMS',
+        'sampler2DMSArray',
+        'isampler2DMSArray',
+        'usampler2DMSArray',
+        'samplerBuffer',
+        'isamplerBuffer',
+        'usamplerBuffer',
+        'samplerShadow',
+
+        // misc texture
```
Comment (mattgodbolt) [inline static/modes/glsl-mode.ts:308]: are these really keywords? These seem like built in functions?

**Review by mattgodbolt — COMMENTED** [review]
A few minor comments but basically LGTM. I don't know if having that gargantuan list of "keywords" is worthwhile, but you know the audience better than me.

File: static/modes/glsl-mode.ts (hunk 5 above)
Comment (spencer-lunarg) [reply static/modes/glsl-mode.ts:308]: well I guess not sure if these counted, like they are reserved names in the GLSL language, just they are functions... happy to remove or group together

File: static/modes/glsl-mode.ts (hunk 4 above)
Comment (spencer-lunarg) [reply static/modes/glsl-mode.ts:163]: yes, can remove some... I was just scanning up-and-down the GLSL list and tried to pull out the C/C++ based keywords

File: examples/glsl/default.glsl (hunk 2 above)
Comment (spencer-lunarg) [reply examples/glsl/default.glsl:12]: nope, just bad editor settings

File: static/modes/glsl-mode.ts (hunk 5 above)
Comment (spencer-lunarg) [reply static/modes/glsl-mode.ts:302]: (I also was kinda surprised by how many keywords GLSL has accumulated, I guess my one counter-argument is for HLSL they also have functions in the keyword list, so it is at least consistent

Comment (mattgodbolt) [reply static/modes/glsl-mode.ts:302]: I'm cool with that then, if you're following what HLSL is doing.

File: static/modes/glsl-mode.ts (hunk 4 above)
Comment (spencer-lunarg) [reply static/modes/glsl-mode.ts:162]: I pulled out more (https://github.com/compiler-explorer/compiler-explorer/pull/6883/commits/9ebf47600640c034abb4781b9398ce1560d24bde), I guess not sure a better way to do a de-duplication check outside what I could grep from `monaco-editor/esm/vs/basic-languages/cpp/cpp`

**Review by mattgodbolt — APPROVED** [review]
LGTM; will merge shortly (unless you let me know otherwise pretty quickly)

**Comment by spencer-lunarg** [comment]
> LGTM; will merge shortly 

awesome, thanks for the review.. I guess what is the step to get `glslang` into the infra so that this can be ran live?

**Comment by mattgodbolt** [comment]
Next step is to get the GLSL compilers installed in our compiler-explorer/infra repository (find the nearest installer you can find there). Once they're installed in our infrastructure, we can the PR and add a `glsl.amazon.properties` that refers to them.

File: etc/config/glsl.defaults.properties [hunk 6]
```diff
@@ -0,0 +1,12 @@
+compilers=&glslang
```
Comment (mattgodbolt) [inline etc/config/glsl.defaults.properties:1]: These defaults aren't used by Compiler Explorer itself, so to get this on the live site we'll need to make a `glsl.amazon.properties`. I'll test locally, merge and add it.

Comment (spencer-lunarg) [reply etc/config/glsl.defaults.properties:1]: > I'll test locally, merge and add it

Thanks!!!

**Pull request merged**